core/src/main/java/org/apache/calcite/sql/SqlDialect.java

```diff
@@ -346,14 +346,7 @@ public RelDataTypeSystem getTypeSystem() {
    * @return Quoted identifier
    */
   public String quoteIdentifier(String val) {
-    if (identifierQuoteString == null) {
-      return val; // quoting is not supported
-    }
-    String val2 =
-        val.replaceAll(
-            identifierEndQuoteString,
-            identifierEscapedQuote);
-    return identifierQuoteString + val2 + identifierEndQuoteString;
+    return quoteIdentifier(new StringBuilder(), val).toString();
```

Comment:
just a side note: jdk has some sort of string concatenation optimizations after JDK9. In the future after JDK usage is upgrade, there might be less need to do such rewrite.

Reply:
Fair point.

I did it mainly to remove code (unify the two `quoteIdentifier` methods, and remove the override for `BigQueryDialect`).